core/src/ast/validity.rs

```diff
@@ -0,0 +1,109 @@
+use super::{Path, TypeName};
+
+#[cfg_attr(feature = "displaydoc", derive(displaydoc::Display))]
+#[derive(Debug, Clone)]
+pub enum ValidityError {
```

Comment:
I guess eventually, we'd just need a mapping from this type to `Diagnostic` from https://github.com/brendanzab/codespan (or equivalent) to get fancy formatting for error messages?

I wonder how passing through detailed spans for error details will work, I'm a bit worried that each enum would become quite large.

Reply:
Yes, either codespan or annotate-snippets-rs.

I think we can just store a single span alongside the error to start, but if we really want we can do the full span thing by storing spans on all AST nodes (or storing the underlying syn node, which has spans)